frontend/src/js/components/PageViewer/VirtualScroll.tsx

```diff
@@ -147,30 +185,47 @@ export const VirtualScroll: FC<VirtualScrollProps> = ({
 
   useLayoutEffect(() => {
     if (viewport?.current && focusedFindHighlight) {
-      const topOfHighlightPage = (pageHeight * (focusedFindHighlight.pageNumber - 1));
+      const highlightYPos = focusedFindHighlight.firstSpan?.y || 0;
+
+      const topOfHighlightPage = (pageHeight * (focusedFindHighlight.pageNumber - 1)) + (highlightYPos * scale);    
+      
       viewport.current.scrollTop = topOfHighlightPage;
     }
-  }, [pageHeight, focusedFindHighlight]);
+  }, [pageHeight, focusedFindHighlight, scale]);
+
   useLayoutEffect(() => {
     if (viewport?.current && focusedSearchHighlight) {
-      const topOfHighlightPage = (pageHeight * (focusedSearchHighlight.pageNumber - 1));
+      const highlightYPos = focusedSearchHighlight.firstSpan?.y || 0;
+
+      const topOfHighlightPage = (pageHeight * (focusedSearchHighlight.pageNumber - 1)) + (highlightYPos * scale);
+      
       viewport.current.scrollTop = topOfHighlightPage;
     }
-  }, [pageHeight, focusedSearchHighlight]);
+  }, [pageHeight, focusedSearchHighlight, scale]);
+
+  const getCachedPage = useCallback((pageNumber: number) => {
+    const cachedPage = pageCache.getPage(pageNumber);
+    if (cachedPage.previewContainerSize === pageCache.containerSize){
+      return cachedPage;
+    } else {
+      return pageCache.refreshPreview(pageNumber, cachedPage.preview, containerSize);
+    }
+  }, [pageCache, containerSize]);
```

Comment:
This function checks if the preview for a specific page number in cache is for the relevant container size or not. If not, a refresh preview is triggered

Reply:
Agreed, it's reasonable that a user might zoom in, then zoom out. So having the cache keyed off the number and the zoom level is reasonable!

I also think we should continue to display the page in a lower resolution until the new resolution has rendered, which will prevent the flicker when the page is very large.

Reply:
Agree with both of you. This would be a must have optimisation. So 2 optimisations we need are:

1. update caching logic in order to keep hold of page previews for different scales in the cache rather than replacing the page preview 
2. Display the new preview after the pdf rendering has completed (For this we need to use the RenderTask.promise)

I had a chat with Jo and he agreed that the option 1, could be done in a separate PR. Would that be OK with you @itsibitzi ?

For option 2 @itsibitzi , I think it'd be better to focus on it in a separate PR, unless if you think this is necessary before we merge this PR?

Reply:
I'm happy with both of those to be done in a separate PR